stablehlo/reference/InterpreterApi.h

```diff
@@ -0,0 +1,95 @@
+/* Copyright 2023 The StableHLO Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef STABLEHLO_REFERENCE_INTERPRETERAPI_H
+#define STABLEHLO_REFERENCE_INTERPRETERAPI_H
+
+#include "llvm/Support/Error.h"
+#include "llvm/Support/ErrorOr.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Support/LLVM.h"
+#include "stablehlo/reference/InterpreterValue.h"
+#include "stablehlo/reference/Process.h"
+#include "stablehlo/reference/Scope.h"
+
+namespace mlir {
+namespace stablehlo {
+
+class InterpreterFallback;
+struct InterpreterConfiguration {
+  InterpreterConfiguration()
+      : fallback(std::make_unique<InterpreterFallback>()) {}
+
+  /// If specified, the directory to which StableHLO interpreter tensors will
+  /// be serialized to disk.
+  std::string probeInstrumentationDir = "";
+
+  /// If specified, use this function as the entrypoint function into the model.
+  /// Will otherwise default to “main”.
+  std::string mainFunction = "main";
+
+  /// If specified, use the callback to run on ops which do not have a
+  /// registered kernel.
+  std::unique_ptr<InterpreterFallback> fallback;
+
+  /// If set, optionally dump tensor values to the specified stream.
+  raw_ostream *stream = nullptr;
+};
+
+/// Base interpreter fallback callback functor to run when no registered kernels
+/// are found for a given StableHLO operation.
+class InterpreterFallback {
+ public:
+  llvm::Error operator()(Operation &op, Process *process, Scope &scope);
+
+  virtual ~InterpreterFallback() = default;
+
+  /// Set the user provided interpreter configuration.
+  void setConfig(const InterpreterConfiguration &config) {
+    this->config = &config;
+  }
+
+  /// Set the topmost currently executing function in the module.
+  void setFunction(func::FuncOp function) { currentFunction = function; }
+
+ protected:
+  /// Custom op kernels for any user specified ops not found in the StableHLO
+  /// op dialect or StableHLO interpreter dialect.
+  virtual llvm::Error handleOp(Operation &op, Process *process, Scope &scope);
+
+  /// The user provided interpreter configuration.
+  const InterpreterConfiguration *config;
```

Comment:
I don't like this setup where we have an `InterpreterConfig` and an `InterpreterFallback` which both must point at eachother if a fallback exists.

Can we make this a unidirectional dependency where a Config has a Fallback and that's it. If there's ever a reason for the back-pointer, I think we should modify other APIs to accept a config instead of a fallback.

If we do this refactoring we can:
- Remove the forward declaration. 
- Update InterpreterConfig to take an InterpreterFallback by reference (open to thoughts on this, I'm ok with pointer, just worth noting that under the current setup we throw away a fallback every time we use `StablehloTranslate` since we override it)

Reply:
I agree, it's a bit confusing of a design. I've refactored it to pass through a config instead to `eval` as the default fallback does need some configuration state (so the fallback functor will still take a config, but it doesn't have a chance of being "outdated" by the toplevel caller not correctly calling `setConfig` before the fallback runs).

However, I think there's an opportunity to further clean up some of the current interpreter design here. When changing the interface, I noticed there's a multiple op kernels which take a `Process` and `Scope`. For example, `evalAllReduceOp`, `evalCaseOp`, `evalIfOp` to name a few, which later go on to recursively call back into teh top level `eval`. However, they don't preserve the current top level fallback function.

For example, `evalIfOp` calls back into `eval` for both true/false branches:

```
  return pred.get({}).getBooleanValue()
             ? eval(trueBranch, {}, /*config=*/nullptr, process, &scope)
             : eval(falseBranch, {}, /*config=*/nullptr, process, &scope);
```

But presumably, if either branch had an op which was handled in the top level specified fallback, that wouldn't work. I'm not sure if this is by design, or an omission. If the latter, since `eval` would also need to take an `InterpreterConfiguration` for properly handling interpreter instrumentation, would it be worthwhile to instead encapsulate `Process`, `Scope` and `InterpreterConfiguration` into something like an `InterpreterState` struct?

Reply:
Ah ignore my other comment, I didn't realize these bugs existed prior to your change. This is on us to clean up. I'll make a ticket with this refactoring / bug fixing

Reply:
Created https://github.com/openxla/stablehlo/issues/1820

Thanks for calling this out. Feel free to resolve comment, I'm currently impartial to the refactoring you've mentioned. If we end up needing to refactor all callsites with a new signature for this context we may reconsider this